File: apps/php-sdk/src/Models/AgentOptions.php
<?php

declare(strict_types=1);

namespace Firecrawl\Models;

final class AgentOptions
{
    /**
     * @param list<string>|null          $urls
     * @param array<string, mixed>|null  $schema
     * @param array<string, string>|null $auditMetadata
     */
    private function __construct(
        private readonly ?array $urls = null,
        private readonly ?string $prompt = null,
        private readonly ?array $schema = null,
        private readonly ?string $integration = null,
        private readonly ?int $maxCredits = null,
        private readonly ?bool $strictConstrainToURLs = null,
        private readonly ?string $model = null,
        private readonly ?WebhookConfig $webhook = null,
        /** @var array<string, string>|null */
        private readonly ?array $auditMetadata = null,
    ) {}

    /**
     * @param list<string>|null         $urls
     * @param array<string, mixed>|null $schema
     * @param array<string, string>|null $auditMetadata
     */
    public static function with(
        ?array $urls = null,
        ?string $prompt = null,
        ?array $schema = null,
        ?string $integration = null,
        ?int $maxCredits = null,
        ?bool $strictConstrainToURLs = null,
        ?string $model = null,
        ?WebhookConfig $webhook = null,
        ?array $auditMetadata = null,
    ): self {
        return new self(
            $urls, $prompt, $schema, $integration,
            $maxCredits, $strictConstrainToURLs, $model, $webhook, $auditMetadata,
        );
    }

    /** @return array<string, mixed> */
    public function toArray(): array
    {
        $fields = [
            'urls' => $this->urls,
            'prompt' => $this->prompt,
            'schema' => $this->schema,
            'integration' => $this->integration,
            'maxCredits' => $this->maxCredits,
            'strictConstrainToURLs' => $this->strictConstrainToURLs,
            'model' => $this->model,
            'webhook' => $this->webhook?->toArray(),
            'auditMetadata' => $this->auditMetadata,
        ];

        return array_filter($fields, fn (mixed $v): bool => $v !== null);
    }
}

File: apps/ruby-sdk/lib/firecrawl/models/agent_options.rb
# frozen_string_literal: true

module Firecrawl
  module Models
    # Options for starting an agent task.
    #
    # effort: valid values are "low", "medium", and "high"; sets the reasoning
    # budget for the run (every level runs spark-2).
    class AgentOptions
      FIELDS = %i[
        urls prompt schema integration max_credits
        strict_constrain_to_urls model effort webhook audit_metadata
      ].freeze

      attr_reader(*FIELDS)

      def initialize(**kwargs)
        FIELDS.each { |f| instance_variable_set(:"@#{f}", kwargs[f]) }
        if audit_metadata && !audit_metadata.is_a?(AuditMetadata)
          raise ArgumentError, "audit_metadata must be an AuditMetadata"
        end
        raise ArgumentError, "Agent prompt is required" if prompt.nil? || prompt.empty?
      end

      def to_h
        {
          "urls" => urls,
          "prompt" => prompt,
          "schema" => schema,
          "integration" => integration,
          "maxCredits" => max_credits,
          "strictConstrainToURLs" => strict_constrain_to_urls,
          "model" => model,
          "effort" => effort,
          "webhook" => webhook.is_a?(Hash) ? webhook : webhook&.to_h,
          "auditMetadata" => audit_metadata&.to_h,
        }.compact
      end
    end
  end
end
